### PR TITLE
[Hail][Feature] Add product option to index methods

### PR DESCRIPTION
**Pull request by patrick-schultz**
Add `product` option to `Table.index`, and `MatrixTable.{index_rows, index_cols}`. Supports interval joins. Refactored the index methods to reduce code duplication, and make them more consistent with each other. Only case not supporting `product=True` is annotating columns of a matrix table with an interval keyed table.

### CODE REVIEW

File: hail/python/test/hail/matrixtable/test_matrix_table.py [hunk 1]
```diff
@@ -388,6 +388,14 @@ def test_table_join(self):
         self.assertTrue(ds.union_cols(ds.drop(ds.info))
                         .count_rows(), 346)
 
+    def test_table_product_join(self):
+        left = hl.utils.range_matrix_table(5, 1)
+        right = hl.utils.range_table(5)
+        right = right.select(i=hl.range(right.idx + 1, 5)).explode('i').key_by('i')
+        left = left.annotate_rows(matches=right.index(left.row_key, product=True))
+        rows = left.rows()
+        self.assertTrue(rows.all(rows.matches.length() == rows.row_idx))
```
Comment (tpoterba) [inline hail/python/test/hail/matrixtable/test_matrix_table.py]: Can we add a test case here and in test_table.py that tests the content of the joined array, not just the length? Something like annotating the interval as a value (or adding the interval key to the annotated value in all cases) and checking that

File: hail/python/hail/matrixtable.py [hunk 2]
```diff
@@ -2726,6 +2726,8 @@ def index_rows(self, *exprs) -> 'StructExpression':
         ----------
         exprs : variable-length args of :class:`.Expression`
             Index expressions.
+        product : bool
+            If ``True``, value of expression is array of all matches.
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/matrixtable.py]: I agree product is a bad name. What about `all_matches` or `collect_matches`?

[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/python/hail/matrixtable.py]: Konrad?

[the remark this answers is not in the document]
Comment (konradjk) [reply hail/python/hail/matrixtable.py]: Maybe the simpler `array=True`?

[the remark this answers is not in the document]
Comment (konradjk) [reply hail/python/hail/matrixtable.py]: Hmm, this only happens on intervals, right? Yeah, I like something that gets across `all`

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/matrixtable.py]: Not just intervals. It also makes arrays of all rows with the matching key in the non-interval case.

File: hail/python/hail/table.py [hunk 3]
```diff
@@ -1504,8 +1504,8 @@ def index(self, *exprs, product=False) -> 'Expression':
         ----------
         exprs : variable-length args of :class:`.Expression`
             Index expressions.
-        product : bool
-            If ``True``, value of expression is array of all matches.
+        all_matches : bool
```
Comment (konradjk) [inline hail/python/hail/table.py]: You renamed it to `all_matches`, but in the downstream code it's still `product`

**Comment by marksantcroos** [comment]
Just letting know that this PR also addresses my issue in https://discuss.hail.is/t/annotate-matrixtable-with-count-from-another-table/950.

**Pull request merged**